EIPS/eip-6860.md

```diff
@@ -0,0 +1,241 @@
+author: Qi Zhou (@qizhou), Chao Pi (@pichaoqkc), Sam Wilson (@SamWilsn)
+discussions-to: https://ethereum-magicians.org/t/eip-4804-web3-url-to-evm-call-message-translation/8300
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-09-29
+requires: 137
+---
+
+## Abstract
+
+This standard translates an RFC 2396 URI like `web3://uniswap.eth/` to an EVM message such as:
+
+```
+EVMMessage {
+   To: 0xaabbccddee.... // where uniswap.eth's address registered at ENS
+   Calldata: 0x
+   ...
+}
+```
+
+⚠️ This proposal updates [ERC-4804](./eip-4804.md) with minor corrections, clarifications and modifications.
+
+## Motivation
+
+Currently, reading data from Web3 generally relies on a translation done by a Web2 proxy to Web3 blockchain. The translation is mostly done by the proxies such as dApp websites/node service provider/etherscan, which are out of the control of users. The standard here aims to provide a simple way for Web2 users to directly access the content of Web3, especially on-chain Web contents such as SVG/HTML.  Moreover, this standard enables interoperability with other standards already compatible with URIs, like SVG/HTML.
+
+## Specification
+
+This specification only defines read-only (i.e. Solidity's `view` functions) semantics. State modifying functions may be defined as a future extension.
+
+A Web3 URL is an ASCII string in the following form
+
+```
+web3URL = web3Schema [userinfo "@"] contractName [":" chainid] path ["?" query]
+web3Schema = [ "w3://" | "web3://" ]
+contractName = address | [name "." [ subDomain0 "." ... ]] nsProviderSuffix
+path = ["/" method ["/" argument_0 ["/" argument_1 ... ]]]
+argument = [type "!"] value
+query = "attribute_1=value_1 [ "&" attribute_2=value_2 ... ]
+attribute = "returns" | "returnTypes" | other_attribute
+```
+
+where
+
+- **web3Schema** indicates the schema of the URL, which is `web3://` or `w3://` for short.
+- **userinfo** indicates which user is calling the EVM, i.e., "From" field in EVM call message. If not specified, the protocol will use 0x0 as the sender address.
+- **contractName** indicates the contract to be called, i.e., "To" field in the EVM call message. If the **contractName** is an **address**, i.e., 0x + 20-byte-data hex, then "To" will be the address. Otherwise, the name is from a name service. In the second case, **nsProviderSuffix** will be the suffix from name service providers such as "eth", etc. The way to translate the name from a name service to an address will be discussed in later EIPs.
+- **chainid** indicates which chain to resolve **contractName** and call the message. If not specified, the protocol will use the same chain as the name service provider, e.g., 1 for eth. If no name service provider is available, the default chainid is 1.
+- **query** is an optional component containing a sequence of attribute-value pairs separated by "&".
+
+All non-ASCII characters must be encoded with URI percent-encoding (RFC 3986).
```

Comment:
You're right, and even looking deeper into RFC 3986, I find out that they split the reserved chars into 2 groups: 

```
reserved    = gen-delims / sub-delims
gen-delims  = ":" / "/" / "?" / "#" / "[" / "]" / "@"
sub-delims  = "!" / "$" / "&" / "'" / "(" / ")"
                  / "*" / "+" / "," / ";" / "="
```

And also that depending of the component, the list of characters that need to be encoded is different:

Path:
```
path          = path-abempty
...
path-abempty  = *( "/" segment )
segment       = *pchar
pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
```

Host (in case of hostname):
```
reg-name    = *( unreserved / pct-encoded / sub-delims )
```

So here in this example they allow not-encoded ":" and "@" in path but not in host, because they have a special meaning in host.
So they optimize to make encoding mandatory when really necessary. I'll update the text to do it likewise here.

Reply:
Ok I have formalized this by adding : 


> 
> The URL must be encoded with URI percent-encoding (RFC 3968) with rules very similar to the RFC. Per component, the rules are :
> - **userinfo** : Identical to **userinfo** as defined in RFC 3986 Appendix A.
> - **contractName** : Identical to **reg-name** as defined in RFC 3986 Appendix A.
> - **chainid** : Identical to **port** as defined in RFC 3986 Appendix A.
> - **path** : Identical to **path-abempty** as defined in RFC 3986 Appendix A, except the character "!" that needs to be percent-encoded.
> - **query** : Identical to **query** as defined in RFC 3986 Appendix A.

Question : I quite like the full formalization of the structure of the URL in [RFC 3986 Appendix A](https://datatracker.ietf.org/doc/html/rfc3986#appendix-A), I think it could be good to make a similar full formalization in an Appendix. What do you think?

Reply:
Thinking about the "!" character in **path** : If we assume that in the future we will not support domain name services that allows "!" as part of domain names (likely I guess), then we can skip this exception and be identical to the RFC ("!" are not required to be encoded in **path**).
I guess that's the way to go.

Reply:
> Ok I have formalized this by adding :
> 
> > The URL must be encoded with URI percent-encoding (RFC 3968) with rules very similar to the RFC. Per component, the rules are :
> > 
> > * **userinfo** : Identical to **userinfo** as defined in RFC 3986 Appendix A.
> > * **contractName** : Identical to **reg-name** as defined in RFC 3986 Appendix A.
> > * **chainid** : Identical to **port** as defined in RFC 3986 Appendix A.
> > * **path** : Identical to **path-abempty** as defined in RFC 3986 Appendix A, except the character "!" that needs to be percent-encoded when not used as type/value separator in an **argument**.
> > * **query** : Identical to **query** as defined in RFC 3986 Appendix A.
> 
> Question : I quite like the full formalization of the structure of the URL in [RFC 3986 Appendix A](https://datatracker.ietf.org/doc/html/rfc3986#appendix-A), I think it could be good to make a similar full formalization in an Appendix. What do you think?

Yes, I think we can make a similar structure to formalize the details for each component.

Reply:
> Thinking about the "!" character in **path** : If we assume that in the future we will not support domain name services that allows "!" as part of domain names (likely I guess), then we can skip this exception and be identical to the RFC ("!" are not required to be encoded in **path**). I guess that's the way to go.
> 
> I have edited:
> 
> > The URL must be encoded with URI percent-encoding (RFC 3968) with rules very similar to the RFC: Component **userinfo** follows **userinfo** encoding as defined in RFC 3986 Appendix A, component **contractName** follows **reg-name**, **chainid** follows **port**, **path** follows **path-abempty**, and **query** follows **query**.

I guess if the address contains `!` we can still use `address!abc%33.eth` to support it?

Reply:
> Yes, I think we can make a similar structure to formalize the details for each component.

Ok I will do that

> I guess if the address contains ! we can still use address!abc%33.eth to support it?

Yes; to be more precise (I think I was a bit confusing) : 

- If we choose to be similar to the RFC, that means that the ``!`` type/value delimiter can be either written as ``!`` or ``%21``. The problem is then : we cannot know if this is part of the domain name or not.. For example, if we have ``address%21abc%21.xyz``, then we cannot know if this is the ``abc!.xyz`` address (detected with the ``<type>!<value>`` syntax), or the ``address!abc!.xyz`` address (detected with auto-detection part 5 as an address)
- The "proper" way to handle that is to require the ``!`` type/value delimiter to always be written as ``!``, not ``%21``. It then makes it a tiny bit of difference with the RFC.

I will do the Appendix with the formal definition, and let's continue from there :)

Reply:
I see.  If we put `!` as type/value delimiter defined in ABNF, then we should never use `%21%`.  I am not sure what is the specific difference with the RFC? 

That said, we could clarify the address part as below (feel free to comment)
- `address!vitalik.eth` is a valid address to lookup `vitalik` from ENS;
- `address%21vitalik.eth` is invalid as it cannot be recognized as any argument;
- `address!%21vitalik.eth` is a valid address to lookup `!vitalik` from ENS;
- `address!vitalik%2Eeth` is invalid as it cannot be recognized as a domain name with a TLD;
- `address!vitalik%2E.eth` is a valid address to lookup `vitalik.` from ENS;
- `address!!vitalik.eth` can be valid or invalid (any thoughts?)

Reply:
I agree that ``domainName`` stricter than ``reg-name`` without reason (we could break with some upcoming domain name services), I have modified it to be identical to RFC 3986 ; and indeed already you have made domain name resolution deferred in upcoming EIPs (such as 6821) ; since 6821 is lower than 6860 I have linked 6821 for ENS and kept the mention "will be discussed in later EIPs for other name services.".

I agree on case 1/ and case 2/ : the ``!`` type/value delimiter is in the ABNF so it's clear ``%21`` is not authorized as a delimiter.
Case 2/ and 6/ : I agree
Case 4/ and 5/ : I would disagree on that; the general workflow would be : 1/ extract ``<value>`` from ``<type>!<value>``, 2/ percent-decode ``<value>`` and then 3/ give ``<percent-decoded-value>`` to the domain name resolver code to do his work -- these would be 2 different layers. (Trying ``https://google%2Ecom`` on a browser, it does percent-decode it into ``google.com`` before doing resolution). So I would think : 
* `address!vitalik%2Eeth` is a valid address  to lookup ``vitalik.eth`` from ENS
* `address!vitalik%2E.eth` is invalid as it would lookup ``vitalik..eth`` from ENS

What do you think?

My modifications are at : 
https://github.com/ethereum/ERCs/compare/master...nand2:ERCs:patch-1

Reply:
Sounds good to me.  I just create the PR with the patch here https://github.com/ethereum/ERCs/pull/84